prophyc/generators/base.py

```diff
@@ -0,0 +1,119 @@
+import os
+from prophyc import model
+
+
+class GenerateError(Exception):
+    pass
+
+
+def _write_file(file_path, string):
+    with open(file_path, "w") as f:
+        f.write(string)
+
+
+def _make_path(output_dir, base_name, extension):
+    assert extension.startswith(".")
+    assert os.path.isdir(output_dir), "Output directory %s doesn't exist." % output_dir
+    return os.path.join(output_dir, base_name + extension)
+
+
+class GeneratorAbc(object):
+    """
+        Generator class implements one language that model is translated to.
+        Generator needs to map at least one translator to its file extension in top_level_translators.
+        And to implement check_nodes method.
+    """
+    top_level_translators = {}
+
+    def check_nodes(self, nodes):
+        """
+            Before a model is translated to given language it needs to be checked if it conforms to given
+            language requirements. It's a place to raise exceptions if needed.
+        """
+
+
+class GeneratorBase(GeneratorAbc):
+    def __init__(self, outpu_directory="."):
+        self.output_dir = outpu_directory
+
+    def serialize(self, nodes, base_name):
+        self.check_nodes(nodes)
+
+        for file_path, translator in self.prepare_translators(base_name):
+            file_content = translator(nodes, base_name)
+            _write_file(file_path, file_content)
+
+    def prepare_translators(self, base_name):
+        for extension, translator_class in self.top_level_translators.items():
+            file_path = _make_path(self.output_dir, base_name, extension)
+            yield file_path, translator_class()
+
+
+class BlockTranslatorBase(object):
```

Comment:
Just a pet peeve - `BlockTranslatorBase` seems a bit confusing to me for following reasons:
1. The core name of this entity is Translator, because it translates from nodes to text.
2. We don't have any other translator than a "Block" one
3. The name starts with a detail ("Block") instead of starting with a core name of mechanism
4. Names of derived translators don't start with "Block"
5. Generator doesn't use any prefix ("Block" or otherwise).
So, according to Occam's Razor, I'd cut word "Block" away - and get an easy to remeber set of names:
- GeneratorAbc, GeneratorBase, Generator<blablabla>
- TranslatorAbc, TranslatorBase, Translator<blablabla>